File: patterns/concurrency/producer-consumer.py
"""
Producer-Consumer Pattern - Python

Coordinates threads/processes where producers generate data and
consumers process that data, typically using a shared buffer/queue.
"""

import queue
import threading
import time


class Producer(threading.Thread):
    def __init__(self, q, name):
        super().__init__()
        self.q = q
        self.name = name
    
    def run(self):
        for i in range(5):
            item = f"Item {i+1}"
            self.q.put(item)
            print(f"{self.name} produced: {item}")
            time.sleep(0.1)


class Consumer(threading.Thread):
    def __init__(self, q, name):
        super().__init__()
        self.q = q
        self.name = name
    
    def run(self):
        while True:
            try:
                item = self.q.get(timeout=1)
                print(f"{self.name} consumed: {item}")
                self.q.task_done()
            except queue.Empty:
                break


# Example usage
if __name__ == "__main__":
    print("=== Producer-Consumer Pattern ===\n")
    
    q = queue.Queue(maxsize=5)
    
    producer = Producer(q, "Producer1")
    consumer = Consumer(q, "Consumer1")
    
    producer.start()
    consumer.start()
    
    producer.join()
    q.join()
    consumer.join()

File: patterns/concurrency/producer-consumer.js
/**
 * Producer-Consumer Pattern - JavaScript
 * 
 * Coordinates threads/processes where producers generate data and
 * consumers process that data, typically using a shared buffer/queue.
 */

class Queue {
    constructor(maxSize = 10) {
        this.items = [];
        this.maxSize = maxSize;
    }
    
    enqueue(item) {
        if (this.items.length >= this.maxSize) {
            throw new Error('Queue is full');
        }
        this.items.push(item);
    }
    
    dequeue() {
        if (this.items.length === 0) {
            return null;
        }
        return this.items.shift();
    }
    
    isEmpty() {
        return this.items.length === 0;
    }
    
    isFull() {
        return this.items.length >= this.maxSize;
    }
}

class Producer {
    constructor(queue, name) {
        this.queue = queue;
        this.name = name;
    }
    
    produce(item) {
        if (!this.queue.isFull()) {
            this.queue.enqueue(item);
            console.log(`${this.name} produced: ${item}`);
            return true;
        }
        console.log(`${this.name} - Queue full, waiting...`);
        return false;
    }
}

class Consumer {
    constructor(queue, name) {
        this.queue = queue;
        this.name = name;
    }
    
    consume() {
        if (!this.queue.isEmpty()) {
            const item = this.queue.dequeue();
            console.log(`${this.name} consumed: ${item}`);
            return item;
        }
        console.log(`${this.name} - Queue empty, waiting...`);
        return null;
    }
}

// Example usage
console.log('=== Producer-Consumer Pattern ===\n');

const queue = new Queue(5);
const producer = new Producer(queue, 'Producer1');
const consumer = new Consumer(queue, 'Consumer1');

// Simulate production and consumption
producer.produce('Item 1');
producer.produce('Item 2');
producer.produce('Item 3');

consumer.consume();
consumer.consume();
producer.produce('Item 4');
consumer.consume();

module.exports = { Queue, Producer, Consumer };

